src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -1,53 +1,277 @@
+}
+
+/** @returns a Comparator that checks whether a test value matches any of the provided values */
+export function anyOf(...values: Value[]): Comparator {
+  return got => {
+    const failed: Array<string> = [];
+    for (const e of values) {
+      const d = diff(got, e);
+      if (d.matched) {
+        return d;
+      }
+      failed.push(d.expected);
+    }
+    return { matched: false, got: got.toString(), expected: failed.join(' or ') };
+  };
+}
+
+// Helper for converting Values to Comparators.
+function toComparator(input: Value | Comparator): Comparator {
+  if ((input as Value).type !== undefined) {
+    return got => diff(got, input as Value);
+  }
+  return input as Comparator;
+}
 
-export type Case<T extends NumberType> = {
-  input: NumberRepr<T>;
-  expected: Array<NumberRepr<T>>;
+/** Case is a single builtin test case. */
+export type Case = {
+  // The input value to the builtin
+  input: Value | Array<Value>;
+  // The expected value, or comparator
+  expected: Value | Comparator;
 };
 
-export function runShaderTestImpl<F extends NumberType>(
+/** CaseList is a list of Cases */
+export type CaseList = Array<Case>;
+
+/** Configuration for running a builtin test */
+export type Config = {
+  // Where the input values are read from
+  storageClass: 'uniform' | 'storage_r' | 'storage_rw';
+  // If defined, scalar test cases will be packed into vectors of the given width
+  vectorize?: number;
+};
+
+// Helper for returning the WGSL storage type for the given Type.
+function storageType(ty: Type): Type {
+  if (ty instanceof ScalarType) {
+    if (ty.kind === 'bool') {
+      return TypeU32;
+    }
+  }
+  if (ty instanceof VectorType) {
+    return TypeVec(ty.width, storageType(ty.elementType) as ScalarType);
+  }
+  return ty;
+}
+
+// Helper for converting a value of the type 'ty' from the storage type.
+function fromStorage(ty: Type, expr: string): string {
+  if (ty instanceof ScalarType) {
+    if (ty.kind === 'bool') {
+      return `${expr} != 0u`;
+    }
+  }
+  if (ty instanceof VectorType) {
+    if (ty.elementType.kind === 'bool') {
+      return `${expr} != vec${ty.width}<u32>(0u)`;
+    }
+  }
+  return expr;
+}
+
+// Helper for converting a value of the type 'ty' to the storage type.
+function toStorage(ty: Type, expr: string): string {
+  if (ty instanceof ScalarType) {
+    if (ty.kind === 'bool') {
+      return `select(0u, 1u, ${expr})`;
+    }
+  }
+  if (ty instanceof VectorType) {
+    if (ty.elementType.kind === 'bool') {
+      return `select(vec${ty.width}<u32>(0u), vec${ty.width}<u32>(0u), ${expr})`;
+    }
+  }
+  return expr;
+}
+
+/**
+ * Runs a builtin test
+ * @param t the GPUTest
+ * @param builtin the builtin being tested
+ * @param parameterTypes the list of builtin parameter types
+ * @param returnType the return type for the builtin overload
+ * @param cfg test configuration values
+ * @param cases list of test cases
+ */
+export function run(
```

Comment:
This makes the common case really nice.

One thing this doesn't handle is, for example, what I've done with the isNan tests https://github.com/gpuweb/cts/pull/800/files#diff-c4a58312f4e5e245645aaccd1702c362a848dd0e0096ee2bc9f79e903081ea9bR79 

where I construct specific bit patterns that map to NaNs, and handle them as Uints on the host side, but they are read and manipulated as f32 on the device side.

I think that can be handled with a special hybrid of F32 and U32.  I can do that as followup, because it's very niche.

Reply:
Does `F32Bits` not give you what you need?